File: 0005-LongestPalindromicSubstring/LongestPalindromicSubstring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        t = "^#" + "#".join(list(s)) + "#$"
        P = [0] * t.__len__()
        center = 0; right = 0
        for i in range(2, t.__len__()):
            mirror = 2*center - i
            if i - 1 - P[i] >= 0 and i + 1 + P[i] < len(t) and i < right:
                P[i] = min(right - i, P[mirror])
            while P[i - 1 - P[i]] == P[i + 1 + P[i]]:
                P[i] = P[i] + 1
            if i + P[i] > right:
                center = i; right = i + P[i]
        maxLen = 0; centerIndex = 0
        for i in range(t.__len__()):
            if P[i] > maxLen:
                maxLen = P[i]
                centerIndex = i
        start = (centerIndex - maxLen) // 2
        return s[start: start + maxLen]

File: 0005-LongestPalindromicSubstring/LongestPalindromicSubstring.java
import java.util.Arrays;

public class LongestPalindromicSubstring {
    public class Solution {
        public String longestPalindrome(String s) {
            // Chuẩn hóa chuỗi
            String t = "^#" + String.join("#", s.split("")) + "#$";
            int n = t.length();
            int[] P = new int[n];
            Arrays.fill(P, 0);

            int center = 0, right = 0;

            for(int i = 1; i < n - 1; i++){
                int mirror = 2 * center - i;

                if(i < right){
                    P[i] = Math.min(right - i, P[mirror]);
                }

                // Mở rộng quanh i
                while(i - 1 - P[i] >= 0 && i + 1 + P[i] < n && t.charAt(i - 1 - P[i]) == t.charAt(i + 1 + P[i])){
                    P[i]++;
                }

                // Cập nhật center và right nếu mở rộng xa hơn
                if(i + P[i] > right){
                    center = i;
                    right = i + P[i];
                }
            }

            // Tìm palindrome dài nhất
            int maxLength = 0, centerIndex = 0;
            for(int i = 1; i < n - 1; i++){
                if(P[i] > maxLength){
                    maxLength = P[i];
                    centerIndex = i;
                }
            }

            int start = (centerIndex - maxLength) / 2;
            return s.substring(start, start + maxLength);
        }
    }
}
